File: android/app/src/main/java/gg/yappy/app/ui/media/MediaViewer.kt
package gg.yappy.app.ui.media

import android.content.ClipData
import android.content.Context
import android.content.Intent
import android.net.Uri
import androidx.compose.animation.core.animateFloatAsState
import androidx.compose.foundation.background
import androidx.compose.foundation.gestures.detectTapGestures
import androidx.compose.foundation.gestures.awaitEachGesture
import androidx.compose.foundation.gestures.awaitFirstDown
import androidx.compose.foundation.gestures.calculatePan
import androidx.compose.foundation.gestures.calculateZoom
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.statusBarsPadding
import androidx.compose.foundation.layout.navigationBarsPadding
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.pager.HorizontalPager
import androidx.compose.foundation.pager.rememberPagerState
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.rounded.Check
import androidx.compose.material.icons.rounded.Close
import androidx.compose.material.icons.rounded.Download
import androidx.compose.material.icons.rounded.Share
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableFloatStateOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.alpha
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.graphicsLayer
import androidx.compose.ui.input.pointer.pointerInput
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.semantics.Role
import androidx.compose.ui.semantics.role
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import androidx.compose.ui.window.Dialog
import androidx.compose.ui.window.DialogProperties
import androidx.core.content.FileProvider
import coil.compose.AsyncImage
import gg.yappy.app.AppContainer
import gg.yappy.app.BuildConfig
import gg.yappy.app.LocalContainer
import gg.yappy.app.ui.components.softClickable
import java.io.File
import java.io.OutputStream
import kotlin.math.abs
import kotlin.math.roundToInt
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import okhttp3.HttpUrl.Companion.toHttpUrlOrNull
import okhttp3.Request

private enum class SaveState { Idle, Saving, Saved }

/** A display name the gallery and the share sheet can both live with. */
private fun ViewerItem.safeName(): String =
    filename?.substringAfterLast('/')?.takeIf { it.isNotBlank() }
        ?: "yappy-${System.currentTimeMillis()}.jpg"

private fun mimeFor(name: String): String = when (name.substringAfterLast('.', "").lowercase()) {
    "png" -> "image/png"
    "gif" -> "image/gif"
    "webp" -> "image/webp"
    else -> "image/jpeg"
}

/**
 * Pull the bytes through the app's own client.
 *
 * Attachments live in a private bucket behind an authorised route, so a bare
 * `java.net.URL` got a 401 and a zero-byte "photo". This mirrors what the Coil
 * loader does for every image on screen: the API's OkHttp client (shared pool,
 * shared timeouts), the in-memory token — never a DataStore read on a network
 * thread — and the header only for our own hosts, so a Tenor GIF or a bot's
 * icon never sees the session.
 */
private fun fetchTo(container: AppContainer, url: String, output: OutputStream): Boolean {
    val parsed = url.toHttpUrlOrNull() ?: return false
    val apiHosts = listOfNotNull(
        BuildConfig.API_URL.toHttpUrlOrNull()?.host,
        BuildConfig.API_URL_ALT.takeIf { it.isNotBlank() }?.toHttpUrlOrNull()?.host,
    )
    val ours = parsed.host in apiHosts
    // The server names itself "localhost"; from inside the emulator that is a
    // different machine entirely. Same rewrite the image loader applies.
    val target = if (BuildConfig.DEBUG && (parsed.host == "localhost" || parsed.host == "127.0.0.1")) {
        parsed.newBuilder().host("10.0.2.2").build()
    } else {
        parsed
    }
    val request = Request.Builder().url(target).apply {
        if (ours) container.session.cachedAccess?.let { header("Authorization", "Bearer $it") }
    }.build()
    container.api.http.newCall(request).execute().use { response ->
        if (!response.isSuccessful) return false
        val body = response.body ?: return false
        body.byteStream().use { input -> input.copyTo(output) }
    }
    return true
}

/**
 * Download the bytes and register them with MediaStore under Pictures/yappy.
 * API 29+ only (checked by the caller): scoped storage means no permission,
 * and IS_PENDING keeps half-written files out of the gallery.
 */
private suspend fun saveToGallery(
    context: Context,
    container: AppContainer,
    item: ViewerItem,
): Boolean = withContext(Dispatchers.IO) {
    runCatching {
        val name = item.safeName()
        val values = android.content.ContentValues().apply {
            put(android.provider.MediaStore.Images.Media.DISPLAY_NAME, name)
            put(android.provider.MediaStore.Images.Media.MIME_TYPE, mimeFor(name))
            put(android.provider.MediaStore.Images.Media.RELATIVE_PATH, "Pictures/yappy")
            put(android.provider.MediaStore.Images.Media.IS_PENDING, 1)
        }
        val resolver = context.contentResolver
        val uri = resolver.insert(
            android.provider.MediaStore.Images.Media.EXTERNAL_CONTENT_URI,
            values,
        ) ?: return@runCatching false

        val ok = resolver.openOutputStream(uri)?.use { output -> fetchTo(container, item.url, output) }
            ?: false
        if (!ok) {
            // Never leave a pending, empty row behind: it would sit in the
            // gallery's database as a photo that never appears.
            resolver.delete(uri, null, null)
            return@runCatching false
        }
        values.clear()
        values.put(android.provider.MediaStore.Images.Media.IS_PENDING, 0)
        resolver.update(uri, values, null, null)
        true
    }.getOrDefault(false)
}

/**
 * Stage a copy under `cache/shared/` and hand out a content URI for it.
 *
 * Sharing the *bytes* rather than the link: the link only resolves for a
 * signed-in member, so pasted into another app it was a URL that opened to a
 * 401 for everyone it was sent to. The FileProvider authority is
 * `<applicationId>.files` with `shared/` as its cache path; if the provider is
 * not declared this returns null and the caller falls back to the link.
 */
private suspend fun stageForShare(
    context: Context,
    container: AppContainer,
    item: ViewerItem,
): Uri? = withContext(Dispatchers.IO) {
    runCatching {
        val dir = File(context.cacheDir, "shared").apply { mkdirs() }
        val file = File(dir, item.safeName())
        val ok = file.outputStream().use { output -> fetchTo(container, item.url, output) }
        if (!ok) {
            file.delete()
            return@runCatching null
        }
        FileProvider.getUriForFile(context, "${context.packageName}.files", file)
    }.getOrNull()
}

/**
 * One image in the viewer, plus the context needed to caption and share it.
 *
 * A plain data class rather than passing `Message` around: the viewer opens
 * from the timeline *and* from the media wall, and those carry different
 * shapes. Flattening at the call site keeps the viewer from knowing about
 * either.
 */
data class ViewerItem(
    val url: String,
    val caption: String? = null,
    val senderName: String? = null,
    val filename: String? = null,
)

/**
 * Full-screen media.
 *
 * A `Dialog` rather than a nav destination on purpose: it opens from the
 * timeline, the media wall and (later) profiles, and routing it would mean
 * every one of those needing a route argument big enough to carry a list of
 * images. As an overlay it composes over whatever is already on screen and the
 * back stack is untouched.
 *
 * Gestures follow what every photo viewer does, because this is the one place
 * where being unsurprising beats being distinctive:
 *   · pinch to zoom, drag to pan while zoomed
 *   · double-tap toggles 1× and 2.5×
 *   · drag down at 1× dismisses, with the background fading as you go
 */
@Composable
fun MediaViewer(
    items: List<ViewerItem>,
    initialIndex: Int,
    onDismiss: () -> Unit,
) {
    if (items.isEmpty()) return
    val context = LocalContext.current
    val container = LocalContainer.current
    val scope = rememberCoroutineScope()

    val pagerState = rememberPagerState(
        initialPage = initialIndex.coerceIn(0, items.lastIndex),
        pageCount = { items.size },
    )

    // How far the current page has been dragged down, in pixels. Drives both
    // the dismiss threshold and the scrim's opacity, so the gesture explains
    // itself as it happens.
    var dragY by remember { mutableFloatStateOf(0f) }
    val dismissProgress = (abs(dragY) / 600f).coerceIn(0f, 1f)

    // Hoisted out of the image so the pager can stand down while one is zoomed.
    var zoomed by remember { mutableStateOf(false) }

    Dialog(
        onDismissRequest = onDismiss,
        properties = DialogProperties(
            usePlatformDefaultWidth = false,
            dismissOnClickOutside = false,
            // Without this the dialog window is inset from the system bars, so
            // the black fill stops short *and* the insets padding below adds a
            // second gap that pushes the caption off screen. A full-screen
            // viewer has to own the whole display and do its own insetting.
            decorFitsSystemWindows = false,
        ),
    ) {
        Box(
            Modifier
                .fillMaxSize()
                .background(Color.Black.copy(alpha = 1f - dismissProgress * 0.7f)),
        ) {
            HorizontalPager(
                state = pagerState,
                // Paging is disabled while an image is zoomed in, otherwise
                // panning left on a zoomed photo flips to the next one.
                userScrollEnabled = !zoomed,
                modifier = Modifier.fillMaxSize(),
            ) { page ->
                ZoomableImage(
                    url = items[page].url,
                    contentDescription = items[page].filename,
                    onDragged = { dragY = it },
                    onZoomChanged = { zoomed = it },
                    onDismiss = onDismiss,
                )
            }

            // ── Chrome ───────────────────────────────────────────────────────
            Row(
                Modifier
                    .fillMaxWidth()
                    // Same reasoning as the caption's bottom padding.
                    .statusBarsPadding()
                    .padding(start = 12.dp, end = 12.dp, top = 34.dp, bottom = 8.dp),
                verticalAlignment = Alignment.CenterVertically,
            ) {
                ViewerButton(Icons.Rounded.Close, "Close", onClick = onDismiss)
                Spacer(Modifier.weight(1f))
                if (items.size > 1) {
                    Text(
                        "${pagerState.currentPage + 1} / ${items.size}",
                        style = MaterialTheme.typography.labelLarge,
                        color = Color.White.copy(alpha = 0.85f),
                    )
                    Spacer(Modifier.weight(1f))
                }
                var sharing by remember { mutableStateOf(false) }
                ViewerButton(Icons.Rounded.Share, "Share", busy = sharing) {
                    if (sharing) return@ViewerButton
                    val item = items[pagerState.currentPage]
                    sharing = true
                    scope.launch {
                        val staged = stageForShare(context, container, item)
                        sharing = false
                        val send = Intent(Intent.ACTION_SEND).apply {
                            if (staged != null) {
                                type = mimeFor(item.safeName())
                                putExtra(Intent.EXTRA_STREAM, staged)
                                // ClipData is what actually carries the grant
                                // to the chooser's target on modern Android;
                                // the flag alone is honoured only for the
                                // intent's own data URI.
                                clipData = ClipData.newRawUri(null, staged)
                                addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION)
                            } else {
                                // Could not stage the bytes (offline, or no
                                // provider): the link is still the honest
                                // fallback — it resolves for anyone who may
                                // see the photo.
                                type = "text/plain"
                                putExtra(Intent.EXTRA_TEXT, item.url)
                            }
                        }
                        context.startActivity(Intent.createChooser(send, "Share"))
                    }
                }
                Spacer(Modifier.width(8.dp))
                var saveState by remember { mutableStateOf(SaveState.Idle) }
                ViewerButton(
                    when (saveState) {
                        SaveState.Saved -> Icons.Rounded.Check
                        else -> Icons.Rounded.Download
                    },
                    "Save",
                    busy = saveState == SaveState.Saving,
                ) {
                    if (saveState == SaveState.Saving) return@ViewerButton
                    val item = items[pagerState.currentPage]
                    if (android.os.Build.VERSION.SDK_INT >= 29) {
                        // A real save: the bytes land in Pictures/yappy via
                        // MediaStore, no permission needed on 10+. Below 10
                        // that path wants WRITE_EXTERNAL_STORAGE, which is not
                        // worth carrying for a shrinking sliver of devices —
                        // they get the old view-in-browser behaviour.
                        saveState = SaveState.Saving
                        scope.launch {
                            val ok = saveToGallery(context, container, item)
                            saveState = if (ok) SaveState.Saved else SaveState.Idle
                            if (!ok) {
                                android.widget.Toast
                                    .makeText(context, "Couldn't save that", android.widget.Toast.LENGTH_SHORT)
                                    .show()
                            }
                        }
                    } else {
                        context.startActivity(
                            Intent(Intent.ACTION_VIEW, android.net.Uri.parse(item.url)),
                        )
                    }
                }
            }

            // ── Caption ──────────────────────────────────────────────────────
            val current = items.getOrNull(pagerState.currentPage)
            val caption = current?.caption?.takeIf { it.isNotBlank() }
            if (caption != null || current?.senderName != null) {
                Column(
                    Modifier
                        .align(Alignment.BottomStart)
                        .fillMaxWidth()
                        .background(Color.Black.copy(alpha = 0.45f))
                        // A Dialog is its own window and does not reliably
                        // receive the activity's insets, so `navigationBarsPadding`
                        // can resolve to zero here and drop the caption behind
                        // the gesture bar. The fixed floor is the belt: it is
                        // roughly a gesture bar's height, and on a device that
                        // does dispatch insets the two simply add to a slightly
                        // roomier caption rather than a broken one.
                        .navigationBarsPadding()
                        .padding(start = 18.dp, end = 18.dp, top = 14.dp, bottom = 30.dp),
                    verticalArrangement = Arrangement.spacedBy(3.dp),
                ) {
                    current?.senderName?.let {
                        Text(
                            it,
                            style = MaterialTheme.typography.labelMedium,
                            color = Color.White.copy(alpha = 0.7f),
                        )
                    }
                    caption?.let {
                        Text(
                            it,
                            style = MaterialTheme.typography.bodyLarge,
                            color = Color.White,
                            maxLines = 4,
                            overflow = TextOverflow.Ellipsis,
                        )
                    }
                }
            }
        }
    }
}

/**
 * The 40dp disc is the picture; the 48dp box around it is the target. Over a
 * photo the chrome should stay small, but the thumb landing on it should not
 * have to be precise.
 */
@Composable
private fun ViewerButton(
    icon: androidx.compose.ui.graphics.vector.ImageVector,
    label: String,
    /** A download in flight: dimmed so a second tap reads as "wait", not "broken". */
    busy: Boolean = false,
    onClick: () -> Unit,
) {
    Box(
        Modifier
            .size(48.dp)
            .semantics { role = Role.Button }
            .softClickable(onClick = onClick),
        contentAlignment = Alignment.Center,
    ) {
        Box(
            Modifier
                .size(40.dp)
                .clip(CircleShape)
                .background(Color.White.copy(alpha = 0.14f))
                .alpha(if (busy) 0.5f else 1f),
            contentAlignment = Alignment.Center,
        ) {
            Icon(icon, label, tint = Color.White, modifier = Modifier.size(20.dp))
        }
    }
}

@Composable
private fun ZoomableImage(
    url: String,
    contentDescription: String?,
    onDragged: (Float) -> Unit,
    onZoomChanged: (Boolean) -> Unit,
    onDismiss: () -> Unit,
) {
    var scale by remember { mutableFloatStateOf(1f) }
    var offsetX by remember { mutableFloatStateOf(0f) }
    var offsetY by remember { mutableFloatStateOf(0f) }
    var dragY by remember { mutableFloatStateOf(0f) }

    // Snapping back is animated so a released pinch settles rather than jumps.
    val animatedScale by animateFloatAsState(scale, label = "zoom")

    Box(
        Modifier
            .fillMaxSize()
            .pointerInput(url) {
                detectTapGestures(
                    onDoubleTap = { tap ->
                        if (scale > 1f) {
                            scale = 1f
                            offsetX = 0f
                            offsetY = 0f
                        } else {
                            scale = 2.5f
                            // Zoom toward the tapped point rather than the
                            // centre, so double-tapping a face centres a face.
                            offsetX = (size.width / 2f - tap.x) * 1.5f
                            offsetY = (size.height / 2f - tap.y) * 1.5f
                        }
                        onZoomChanged(scale > 1f)
                    },
                )
            }
            /**
             * Written with `awaitEachGesture` rather than
             * `detectTransformGestures` because that helper consumes every
             * event it sees — which silently swallows the horizontal drag the
             * pager needs, leaving a photo viewer you cannot swipe through.
             *
             * So consumption is deliberate here: only when this gesture is
             * actually a pinch, a pan on an already-zoomed image, or a vertical
             * drag to dismiss. A plain horizontal drag at 1× is left alone and
             * reaches the pager.
             */
            .pointerInput(url) {
                awaitEachGesture {
                    awaitFirstDown(requireUnconsumed = false)
                    var totalX = 0f
                    var totalY = 0f
                    // Decided once per gesture: flipping mid-drag between
                    // "dismissing" and "paging" feels broken.
                    var claimedVertical = false

                    do {
                        val event = awaitPointerEvent()
                        if (event.changes.any { it.isConsumed } && scale <= 1f && !claimedVertical) break

                        val zoomChange = event.calculateZoom()
                        val pan = event.calculatePan()
                        val pinching = event.changes.count { it.pressed } > 1

                        totalX += pan.x
                        totalY += pan.y

                        when {
                            pinching || scale > 1f -> {
                                val next = (scale * zoomChange).coerceIn(1f, 5f)
                                scale = next
                                if (next > 1f) {
                                    // Clamped to the scaled image's bounds, so
                                    // the picture cannot be dragged off into
                                    // empty space.
                                    val maxX = (size.width * (next - 1)) / 2f
                                    val maxY = (size.height * (next - 1)) / 2f
                                    offsetX = (offsetX + pan.x).coerceIn(-maxX, maxX)
                                    offsetY = (offsetY + pan.y).coerceIn(-maxY, maxY)
                                } else {
                                    offsetX = 0f
                                    offsetY = 0f
                                }
                                onZoomChanged(scale > 1f)
                                event.changes.forEach { it.consume() }
                            }

                            // At 1×, a decisively vertical drag means "put this
                            // away". Anything else belongs to the pager.
                            claimedVertical || abs(totalY) > abs(totalX) * 1.5f && abs(totalY) > 24f -> {
                                claimedVertical = true
                                dragY += pan.y
                                onDragged(dragY)
                                event.changes.forEach { it.consume() }
                                if (dragY > 220f) {
                                    onDismiss()
                                    return@awaitEachGesture
                                }
                            }
                        }
                    } while (event.changes.any { it.pressed })

                    // Released without committing: settle back.
                    if (claimedVertical) {
                        dragY = 0f
                        onDragged(0f)
                    }
                }
            },
        contentAlignment = Alignment.Center,
    ) {
        AsyncImage(
            model = url,
            contentDescription = contentDescription,
            contentScale = ContentScale.Fit,
            modifier = Modifier
                .fillMaxSize()
                .graphicsLayer(
                    scaleX = animatedScale,
                    scaleY = animatedScale,
                    translationX = offsetX,
                    translationY = if (scale > 1f) offsetY else dragY,
                ),
        )
    }
}

/** Reset the drag when a page settles, so a half-swipe does not persist. */
internal fun clampDrag(value: Float): Int = value.roundToInt()

File: ios/Yappy/Features/Media/MediaViewer.swift
import SwiftUI
import UIKit

/// One image in the viewer, plus the context needed to caption and share it.
///
/// A flat struct rather than passing `Message` around: the viewer opens from the
/// timeline *and* from the media wall, and those carry different shapes.
/// Flattening at the call site keeps the viewer from knowing about either.
struct ViewerItem: Identifiable, Hashable {
    let id: String
    let url: String
    var caption: String?
    var senderName: String?
    var filename: String?
}

/// Full-screen media.
///
/// Presented as a cover rather than a navigation destination on purpose: it
/// opens from the timeline, the media wall and (later) profiles, and routing it
/// would mean every one of those needing a route argument big enough to carry a
/// list of images. As a cover it composes over whatever is already on screen and
/// the back stack is untouched.
///
/// Gestures follow what every photo viewer does, because this is the one place
/// where being unsurprising beats being distinctive:
///   · pinch to zoom, drag to pan while zoomed
///   · double-tap toggles 1× and 2.5×
///   · drag down at 1× dismisses, with the background fading as you go
struct MediaViewer: View {
    let items: [ViewerItem]
    let onDismiss: () -> Void

    /// Seeded in `init` rather than corrected in `onAppear`, which runs after
    /// the pager has already laid out and drawn page zero — opening the seventh
    /// photo on a wall showed the first one and then jumped.
    @State private var index: Int

    init(items: [ViewerItem], initialIndex: Int, onDismiss: @escaping () -> Void) {
        self.items = items
        self.onDismiss = onDismiss
        _index = State(initialValue: min(max(initialIndex, 0), max(items.count - 1, 0)))
    }

    /// How far the current page has been dragged down. Drives both the dismiss
    /// threshold and the scrim's opacity, so the gesture explains itself as it
    /// happens.
    @State private var dragY: CGFloat = 0
    /// Hoisted out of the image so paging can stand down while one is zoomed.
    @State private var zoomed = false
    /// The photo is for looking at; the buttons and the caption are for acting
    /// on it. One tap puts the furniture away, another brings it back.
    @State private var chromeHidden = false

    /// Items already written to the photo library this presentation, so the
    /// button can answer with a checkmark instead of silently writing twice.
    @State private var saved: Set<String> = []
    @State private var saving = false
    /// `UIImageWriteToSavedPhotosAlbum` calls back through a target/selector;
    /// held in state so the target outlives the write.
    @State private var saveRelay: SaveRelay?

    /// A `Double` because it only ever feeds `opacity`, and mixing CGFloat into
    /// that expression makes the literal `1` ambiguous.
    private var dismissProgress: Double { min(Double(abs(dragY)) / 600, 1) }

    var body: some View {
        ZStack {
            Color.black
                .opacity(1 - dismissProgress * 0.7)
                .ignoresSafeArea()

            TabView(selection: $index) {
                ForEach(Array(items.enumerated()), id: \.element.id) { position, item in
                    ZoomableImage(
                        url: item.url,
                        onDragged: { dragY = $0 },
                        onZoomChanged: { zoomed = $0 },
                        onTap: { chromeHidden.toggle() },
                        onDismiss: onDismiss
                    )
                    .tag(position)
                }
            }
            .tabViewStyle(.page(indexDisplayMode: .never))
            // Deliberately *not* `.disabled(zoomed)`. That was meant to stop a
            // pan on a zoomed photo flipping to the next one, but `disabled`
            // propagates through the whole subtree — it switched off the pinch,
            // the pan and the double-tap inside the image as well. Zooming in
            // was therefore permanent: nothing left could zoom back out, and
            // since only those gestures clear `zoomed`, the only way out of the
            // viewer was the Close button. `ZoomableImage` already swallows the
            // drag itself once it is zoomed, which is the part that mattered.

            VStack {
                chrome
                Spacer()
                caption
            }
            // Faded rather than removed, so bringing it back is a reappearance
            // in place and not a relayout. Hit-testing goes with it: an
            // invisible Close button that still swallows taps would make the
            // photo unresponsive in exactly the corners people tap.
            .opacity(chromeHidden ? 0 : 1)
            .animation(.easeInOut(duration: 0.2), value: chromeHidden)
            .allowsHitTesting(!chromeHidden)
        }
        .statusBarHidden()
    }

    private var chrome: some View {
        HStack {
            ViewerButton(systemName: "xmark", label: "Close", action: onDismiss)
            Spacer()
            if items.count > 1 {
                Text("\(index + 1) / \(items.count)")
                    .font(YappyFont.labelLarge)
                    .foregroundStyle(.white.opacity(0.85))
                Spacer()
            }
            if let current = items[safe: index] {
                ViewerButton(
                    systemName: saved.contains(current.id) ? "checkmark" : "arrow.down.to.line",
                    label: "Save to Photos"
                ) { save(current) }
                .opacity(saving ? 0.5 : 1)
                .padding(.trailing, 10)

                // Sharing the URL rather than the bytes: the file lives in a
                // private bucket behind an authorised route, so handing another
                // app the link is the honest thing — it will only resolve for
                // someone who may see it.
                ShareLink(item: URL(string: current.url) ?? URL(string: "https://yappy.gg")!) {
                    Image(systemName: "square.and.arrow.up")
                        .font(.system(size: 18, weight: .medium))
                        .foregroundStyle(.white)
                        .frame(width: 40, height: 40)
                        .background(.white.opacity(0.14), in: Circle())
                }
                .buttonStyle(.plain)
            }
        }
        .padding(.horizontal, 12)
        .padding(.top, 8)
    }

    /// Save the picture to the photo library.
    ///
    /// The bytes come from `ImageLoader` — the same authed, cached pipeline
    /// that put the image on screen — so a save is usually a cache hit and
    /// never leaks the session token to a foreign host. Failure is silent by
    /// design: the viewer has no chrome for errors, and the write asks for
    /// add-only photo permission on its own (the purpose string is in the
    /// Info.plist).
    private func save(_ item: ViewerItem) {
        guard !saving, !saved.contains(item.id) else { return }
        saving = true
        Task {
            guard let loaded = await ImageLoader.shared.load(item.url) else {
                saving = false
                return
            }
            // A GIF decodes to an animated UIImage; the photo write wants a
            // still, so take the first frame rather than leaving UIKit to guess.
            let image = loaded.images?.first ?? loaded
            let relay = SaveRelay { ok in
                saving = false
                // Animated so the button's `.symbolEffect(.replace)` has a
                // transaction to ride — without one the arrow just becomes a
                // checkmark between frames.
                if ok { withAnimation { saved.insert(item.id) } }
            }
            saveRelay = relay
            UIImageWriteToSavedPhotosAlbum(
                image,
                relay,
                #selector(SaveRelay.image(_:didFinishSavingWithError:contextInfo:)),
                nil
            )
        }
    }

    @ViewBuilder
    private var caption: some View {
        if let current = items[safe: index],
           current.senderName != nil || !(current.caption ?? "").isEmpty {
            VStack(alignment: .leading, spacing: 3) {
                if let sender = current.senderName {
                    Text(sender)
                        .font(YappyFont.labelMedium)
                        .foregroundStyle(.white.opacity(0.7))
                }
                if let caption = current.caption, !caption.isEmpty {
                    Text(caption)
                        .font(YappyFont.bodyLarge)
                        .foregroundStyle(.white)
                        .lineLimit(4)
                }
            }
            .frame(maxWidth: .infinity, alignment: .leading)
            .padding(.horizontal, 18)
            .padding(.vertical, 14)
            .background(.black.opacity(0.45))
        }
    }
}

/// The completion target `UIImageWriteToSavedPhotosAlbum` requires: an NSObject
/// with exactly this selector. A wrong or missing selector there is a crash,
/// which is why the callback is not optional sugar.
private final class SaveRelay: NSObject {
    private let done: (Bool) -> Void

    init(done: @escaping (Bool) -> Void) {
        self.done = done
    }

    @objc func image(_ image: UIImage, didFinishSavingWithError error: NSError?, contextInfo: UnsafeRawPointer) {
        done(error == nil)
    }
}

private struct ViewerButton: View {
    let systemName: String
    let label: String
    let action: () -> Void

    var body: some View {
        Button(action: action) {
            Image(systemName: systemName)
                .font(.system(size: 18, weight: .medium))
                .foregroundStyle(.white)
                // The arrow becomes a checkmark by replacement rather than by
                // redraw, so a save reads as the same button answering instead
                // of a different button appearing.
                .contentTransition(.symbolEffect(.replace))
                .frame(width: 40, height: 40)
                .background(.white.opacity(0.14), in: Circle())
                .contentShape(Circle())
        }
        .buttonStyle(ViewerPress())
        .accessibilityLabel(label)
    }
}

/// The viewer's press: the same small dip `softPress` gives buttons on the
/// sheet, rebuilt plainly here because these float on a photograph — the neu
/// treatment itself stays home.
private struct ViewerPress: ButtonStyle {
    func makeBody(configuration: Configuration) -> some View {
        configuration.label
            .scaleEffect(configuration.isPressed ? 0.92 : 1)
            .animation(.easeOut(duration: 0.12), value: configuration.isPressed)
    }
}

private struct ZoomableImage: View {
    let url: String
    let onDragged: (CGFloat) -> Void
    let onZoomChanged: (Bool) -> Void
    /// A single tap on the photo — the viewer uses it to put the chrome away.
    let onTap: () -> Void
    let onDismiss: () -> Void

    @State private var scale: CGFloat = 1
    @State private var committedScale: CGFloat = 1
    @State private var offset: CGSize = .zero
    @State private var committedOffset: CGSize = .zero
    @State private var dragY: CGFloat = 0
    /// Past the point where letting go dismisses. Tracked so the tick fires
    /// once per gesture, on the way in, rather than on every frame.
    @State private var armed = false

    /// Far enough that a resting thumb never trips it, close enough that a
    /// deliberate pull does not feel refused.
    private let commit: CGFloat = 220

    var body: some View {
        GeometryReader { geometry in
            RemoteImage(url: url, contentMode: .fit)
                .frame(width: geometry.size.width, height: geometry.size.height)
                .scaleEffect(scale)
                .offset(
                    x: offset.width,
                    y: scale > 1 ? offset.height : dragY
                )
                .gesture(magnify(in: geometry.size))
                .simultaneousGesture(pan(in: geometry.size))
                .onTapGesture(count: 2) { toggleZoom() }
                // Declared after the double-tap on purpose: with both present
                // SwiftUI waits out the double before delivering the single,
                // so hiding the chrome cannot eat the zoom.
                .onTapGesture { onTap() }
        }
    }

    private func magnify(in size: CGSize) -> some Gesture {
        MagnifyGesture()
            .onChanged { value in
                scale = min(max(committedScale * value.magnification, 1), 5)
                onZoomChanged(scale > 1)
            }
            .onEnded { _ in
                committedScale = scale
                if scale <= 1 {
                    // Snapping back is animated so a released pinch settles
                    // rather than jumps.
                    withAnimation(.spring(response: 0.3, dampingFraction: 0.8)) {
                        offset = .zero
                    }
                    committedOffset = .zero
                }
                onZoomChanged(scale > 1)
            }
    }

    /// At 1× a decisively vertical drag means "put this away"; anything else
    /// belongs to the pager. Once zoomed, the same gesture pans the picture and
    /// is clamped to the scaled bounds so it cannot be dragged into empty space.
    private func pan(in size: CGSize) -> some Gesture {
        DragGesture()
            .onChanged { value in
                if scale > 1 {
                    let maxX = size.width * (scale - 1) / 2
                    let maxY = size.height * (scale - 1) / 2
                    offset = CGSize(
                        width: min(max(committedOffset.width + value.translation.width, -maxX), maxX),
                        height: min(max(committedOffset.height + value.translation.height, -maxY), maxY)
                    )
                } else if value.translation.height > 0 {
                    dragY = value.translation.height
                    onDragged(dragY)
                    // The tick at the moment letting go would dismiss — the
                    // gesture says what it will do before it does it.
                    if dragY > commit, !armed {
                        armed = true
                        Haptics.tap()
                    }
                }
            }
            .onEnded { value in
                armed = false
                if scale > 1 {
                    committedOffset = offset
                    return
                }
                // Predicted rather than actual: a flick travels a short
                // distance fast, and it means "away" just as clearly as a
                // long slow pull does.
                if value.predictedEndTranslation.height > commit {
                    // Carry the photo off the bottom before the cover goes,
                    // so the dismissal reads as the throw finishing rather
                    // than the picture blinking out. The dismiss itself waits
                    // for the exit — called cold, it tore the view down on
                    // the first frame of the slide.
                    withAnimation(.easeIn(duration: 0.18)) {
                        dragY = size.height
                        onDragged(size.height)
                    }
                    Task {
                        try? await Task.sleep(for: .milliseconds(180))
                        onDismiss()
                    }
                    return
                }
                // Released without committing: settle back.
                withAnimation(.spring(response: 0.3, dampingFraction: 0.85)) { dragY = 0 }
                onDragged(0)
            }
    }

    private func toggleZoom() {
        withAnimation(.spring(response: 0.3, dampingFraction: 0.8)) {
            if scale > 1 {
                scale = 1
                offset = .zero
            } else {
                scale = 2.5
            }
            committedScale = scale
            committedOffset = offset
        }
        onZoomChanged(scale > 1)
    }
}

extension Array {
    /// Bounds-checked lookup. The viewer indexes by page and a page can be one
    /// past the end for a frame while the list shrinks under it.
    subscript(safe index: Int) -> Element? {
        indices.contains(index) ? self[index] : nil
    }
}
